### PR TITLE
ring_hash_lb: wire up ring size and hash fn config

### PR DESCRIPTION
**Pull request by danielhochman**
**Description**
Adds `minimum_ring_size` and `use_std_hash`. Fixes https://github.com/envoyproxy/envoy/issues/2021.

**TODO**
- [x] update docs in data plane https://github.com/envoyproxy/data-plane-api/pull/277

**API Changes**
Completes https://github.com/envoyproxy/data-plane-api/pull/202.

**Risk Level**
Low

**Testing**
UT

### CODE REVIEW

File: include/envoy/upstream/upstream.h [hunk 1]
```diff
@@ -313,6 +313,11 @@ class ClusterInfo {
   virtual LoadBalancerType lbType() const PURE;
 
   /**
+   * @return configuration for ring hash load balancing, only used if type is set to ring_hash_lb.
+   */
+  virtual const envoy::api::v2::Cluster::RingHashLbConfig lbRingHashConfig() const PURE;
```
Comment (ggreenway) [inline include/envoy/upstream/upstream.h]: To mirror the dataplane-api code, should this return something that corresponds to the oneof lb_config?  I'm not sure what the right way pass a "base class" of a protobuf message is, that can be downcast to the correct type.

If not, this should probably be Optional<>.

File: source/common/json/config_schemas.cc [hunk 2]
```diff
@@ -1436,6 +1436,18 @@ const std::string Json::Schema::CLUSTER_SCHEMA(R"EOF(
         "type" : "string",
         "enum" : ["round_robin", "least_request", "random", "ring_hash", "original_dst_lb"]
       },
+      "ring_hash_lb_config" : {
```
Comment (ggreenway) [inline source/common/json/config_schemas.cc]: Does new config need to go into the old v1 schema?  @htuch ?

Comment (danielhochman) [reply source/common/json/config_schemas.cc]: In this case it's needed for migration purposes.

File: include/envoy/upstream/upstream.h (hunk 1 above)
Comment (danielhochman) [reply include/envoy/upstream/upstream.h]: Probably need Optional<>, yeah. I was initially viewing the oneof as input enforcement in the proto and not extending to the code itself.

Let's see what @htuch has to say.

Comment (mattklein123) [reply include/envoy/upstream/upstream.h]: At the interface level here I would probably just do Optional<>, or return a const pointer which can be nullptr. Either way.

File: source/common/json/config_schemas.cc (hunk 2 above)
Comment (mattklein123) [reply source/common/json/config_schemas.cc]: We should add to v1 docs also.

File: source/common/upstream/ring_hash_lb.cc [hunk 3]
```diff
@@ -102,14 +103,15 @@ void RingHashLoadBalancer::Ring::create(Runtime::Loader& runtime,
     }
   }
 
-  ENVOY_LOG(trace, "ring hash: min_ring_size={} hashes_per_host={}", min_ring_size,
-            hashes_per_host);
+  ENVOY_LOG(info, "ring hash: min_ring_size={} hashes_per_host={}", min_ring_size, hashes_per_host);
   ring_.reserve(hosts.size() * hashes_per_host);
+
+  bool use_std_hash = PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.deprecated_v1(), use_std_hash, true);
```
Comment (mattklein123) [inline source/common/upstream/ring_hash_lb.cc]: const

File: source/common/upstream/ring_hash_lb.h [hunk 4]
```diff
@@ -47,6 +49,7 @@ class RingHashLoadBalancer : public LoadBalancer, Logger::Loggable<Logger::Id::u
   ClusterStats& stats_;
   Runtime::Loader& runtime_;
   Runtime::RandomGenerator& random_;
+  envoy::api::v2::Cluster::RingHashLbConfig config_;
```
Comment (mattklein123) [inline source/common/upstream/ring_hash_lb.h]: const

File: test/common/upstream/cluster_manager_impl_test.cc [hunk 5]
```diff
@@ -345,6 +345,25 @@ TEST_F(ClusterManagerImplTest, SubsetLoadBalancerRestriction) {
       "cluster: cluster type 'original_dst' may not be used with lb_subset_config");
 }
 
+TEST_F(ClusterManagerImplTest, RingHashLoadBalancerInitialization) {
+  const std::string json = R"EOF(
```
Comment (mattklein123) [inline test/common/upstream/cluster_manager_impl_test.cc]: If possible can you do a v2 YAML test also? We have been adding new tests w/ v2 YAML config. See some recent PRs for examples on how to do it. It's pretty easy.

Comment (danielhochman) [reply test/common/upstream/cluster_manager_impl_test.cc]: Added one

File: include/envoy/upstream/upstream.h [hunk 6]
```diff
@@ -312,6 +312,11 @@ class ClusterInfo {
    */
   virtual LoadBalancerType lbType() const PURE;
 
+  /**
+   * @return configuration for ring hash load balancing, only used if type is set to ring_hash_lb.
+   */
+  virtual const Optional<envoy::api::v2::Cluster::RingHashLbConfig> lbRingHashConfig() const PURE;
```
Comment (mattklein123) [inline include/envoy/upstream/upstream.h]: `const Optional<...>&`

File: source/common/upstream/subset_lb.h [hunk 7]
```diff
@@ -21,7 +21,8 @@ class SubsetLoadBalancer : public LoadBalancer, Logger::Loggable<Logger::Id::ups
 public:
   SubsetLoadBalancer(LoadBalancerType lb_type, HostSet& host_set, const HostSet* local_host_set,
                      ClusterStats& stats, Runtime::Loader& runtime,
-                     Runtime::RandomGenerator& random, const LoadBalancerSubsetInfo& subsets);
+                     Runtime::RandomGenerator& random, const LoadBalancerSubsetInfo& subsets,
+                     const Optional<envoy::api::v2::Cluster::RingHashLbConfig> lb_ring_hash_config);
```
Comment (mattklein123) [inline source/common/upstream/subset_lb.h]: `const Optional<...>&`

File: source/common/upstream/ring_hash_lb.cc [hunk 8]
```diff
@@ -102,14 +106,18 @@ void RingHashLoadBalancer::Ring::create(Runtime::Loader& runtime,
...
+  const bool use_std_hash =
+      config.valid()
+          ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value().deprecated_v1(), use_std_hash, true)
+          : true;
   for (const auto& host : hosts) {
     for (uint64_t i = 0; i < hashes_per_host; i++) {
       std::string hash_key(host->address()->asString() + "_" + std::to_string(i));
-      // TODO(danielhochman): convert to HashUtil::xxHash64 when we have a migration strategy.
-      uint64_t hash = std::hash<std::string>()(hash_key);
+      uint64_t hash =
```
Comment (mattklein123) [inline source/common/upstream/ring_hash_lb.cc]: const (same for hash_key)

File: test/common/upstream/cluster_manager_impl_test.cc [hunk 9]
```diff
@@ -532,6 +532,8 @@ TEST_F(ClusterManagerImplTest, InitializeOrder) {
       defaultStaticClusterJson("cds_cluster"),
       clustersJson({defaultStaticClusterJson("cluster_0"), defaultStaticClusterJson("cluster_1")}));
 
+  // EXPECT_CALL(*cluster, foo()).WillOnce(Return(Optional<>{}));
```
Comment (mattklein123) [inline test/common/upstream/cluster_manager_impl_test.cc]: ?

Comment (danielhochman) [reply test/common/upstream/cluster_manager_impl_test.cc]: removed

**Pull request merged**